File: jupyter2/ps/parametric/generic_optimize/go/generic_optimize.go
package main
import (
	"fmt"
	"math"
)

/** begin my answer */

/** begin hidden */
type Domain[T any] interface {
	next() (bool, T)
}

type Arange struct {
	a float64
	b float64
	i int
	n int
	dx float64
}

func arange(a, b float64, n int) Arange {
	return Arange{a, b, 0, n, (b - a) / float64(n - 1)}
}

func (a *Arange) next() (bool, float64) {
	if a.i == a.n {
		return false, 0.0
	} else {
		x := a.a + a.dx * float64(a.i)
		a.i += 1
		return true, x
	}
}

type Arange2d struct {
	x0 float64
	x1 float64
	m int
	y0 float64
	y1 float64
	n int
	i int
	j int
	dx float64
	dy float64
}

func arange2d(x0, x1 float64, m int, y0, y1 float64, n int) Arange2d {
	return Arange2d{x0, x1, m, y0, y1, n, 0, 0, (x1 - x0) / float64(m - 1), (y1 - y0) / float64(n - 1)}
}

func (a *Arange2d) next() (bool, [2]float64) {
	if a.i == a.m {
		return false, [2]float64{0.0, 0.0}
	} else {
		x := a.x0 + a.dx * float64(a.i)
		y := a.y0 + a.dy * float64(a.j)
		if a.j + 1 == a.n {
			a.i += 1
			a.j = 0
		} else {
			a.j += 1
		}
		return true, [2]float64{x, y}
	}
}

func optimize[S any](f func(S) float64, rng Domain[S]) (bool, float64) {
	var miny float64
	ok, x := rng.next()
	i := 0
	for ok {
		y := f(x)
		if i == 0 || y < miny {
			miny = y
		}
		i += 1
		ok, x = rng.next()
	}
	return i > 0, miny
}

/** end hidden */
/** end my answer */
func very_close(x, y float64) bool {
	return math.Abs(x - y) < 1e-6
}

func main() {
	a0 := arange(0, 1, 10000)
	a1 := arange(0, 3, 10000)
	a2 := arange2d(-2, 2, 10000, -2, 2, 10000)
	f0 := func (x float64) float64 { return x * (x - 1) }
	f1 := func (x float64) float64 { return x * (x - 1) * (x - 2) }
	f2 := func (x [2]float64) float64 { return x[0] * x[0] + 2 * x[1] * x[1] + 3 * x[0] * x[1] }
	if ok, y0 := optimize(f0, &a0); !(ok && very_close(y0, -0.25))     { panic("wrong") }
	if ok, y1 := optimize(f1, &a1); !(ok && very_close(y1, -0.384900)) { panic("wrong") }
	if ok, y2 := optimize(f2, &a2); !(ok && very_close(y2, -0.5))      { panic("wrong") }
	fmt.Println("OK")
}

File: jupyter2/ps/parametric/generic_optimize/jl/generic_optimize.jl
### begin my answer

### begin hidden
mutable struct Arange
    a::Float64
    b::Float64
    i::Int
    n::Int
    dx::Float64
end

function arange(a, b, n)
    Arange(a, b, 0, n, (b - a) / (n - 1))
end

function next(a::Arange)
    if a.i == a.n
        nothing
    else
        x = a.a + a.dx * a.i
        a.i += 1
        x
    end
end

mutable struct Arange2d
    x0::Float64
    x1::Float64
    m::Int
    y0::Float64
    y1::Float64
    n::Int
    i::Int
    j::Int
    dx::Float64
    dy::Float64
end

function arange2d(x0, x1, m, y0, y1, n)
    dx = (x1 - x0) / (m - 1)
    dy = (y1 - y0) / (n - 1)
    Arange2d(x0, x1, m, y0, y1, n, 0, 0, dx, dy)
end

function next(a::Arange2d)
    if a.i == a.m
        nothing
    else
        x = a.x0 + a.dx * a.i
        y = a.y0 + a.dy * a.j
        if a.j + 1 == a.n
            a.i += 1
            a.j = 0
        else
            a.j += 1
        end
        (x, y)
    end
end

function optimize(f, rng)
    minx, miny = nothing, nothing
    x = next(rng)
    while x != nothing
        y = f(x)
        if miny == nothing || y < miny
            minx, miny = x, y
        end
        x = next(rng)
    end
    miny
end

### end hidden
### end my answer

very_close(x, y) = abs(x - y) < 1e-6

function main()
    a0 = arange(0, 1, 10000)
    a1 = arange(0, 3, 10000)
    a2 = arange2d(-2, 2, 10000, -2, 2, 1000)
    f0(x) = x * (x - 1)
    f1(x) = x * (x - 1) * (x - 2)
    f2((x, y)) = x * x + 2 * y * y + 3 * x * y
    y0 = optimize(f0, a0)
    y1 = optimize(f1, a1)
    y2 = optimize(f2, a2)
    @assert(very_close(y0, -0.25))
    @assert(very_close(y1, -0.384900))
    @assert(very_close(y2, -0.5))
    println("OK")
end

main()
